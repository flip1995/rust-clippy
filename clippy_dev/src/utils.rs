use core::fmt::{self, Display};
use core::str::FromStr;
use std::env;
use std::fs::{self, OpenOptions};
use std::io::{self, Read as _, Seek as _, SeekFrom, Write};
use std::path::{Path, PathBuf};
use std::process::{self, ExitStatus};

#[cfg(not(windows))]
static CARGO_CLIPPY_EXE: &str = "cargo-clippy";
#[cfg(windows)]
static CARGO_CLIPPY_EXE: &str = "cargo-clippy.exe";

#[cold]
#[track_caller]
fn panic_io(e: &io::Error, action: &str, path: &Path) -> ! {
    panic!("error {action} `{}`: {}", path.display(), *e)
}

/// Wrapper around `std::fs::File` which panics with a path on failure.
pub struct File<'a> {
    pub inner: fs::File,
    pub path: &'a Path,
}
impl<'a> File<'a> {
    /// Opens a file panicking on failure.
    #[track_caller]
    pub fn open(path: &'a (impl AsRef<Path> + ?Sized), options: &mut OpenOptions) -> Self {
        let path = path.as_ref();
        match options.open(path) {
            Ok(inner) => Self { inner, path },
            Err(e) => panic_io(&e, "opening", path),
        }
    }

    /// Opens a file if it exists, panicking on any other failure.
    #[track_caller]
    pub fn open_if_exists(path: &'a (impl AsRef<Path> + ?Sized), options: &mut OpenOptions) -> Option<Self> {
        let path = path.as_ref();
        match options.open(path) {
            Ok(inner) => Some(Self { inner, path }),
            Err(e) if e.kind() == io::ErrorKind::NotFound => None,
            Err(e) => panic_io(&e, "opening", path),
        }
    }

    /// Opens and reads a file into a string, panicking of failure.
    #[track_caller]
    pub fn open_read_to_cleared_string<'dst>(
        path: &'a (impl AsRef<Path> + ?Sized),
        dst: &'dst mut String,
    ) -> &'dst mut String {
        Self::open(path, OpenOptions::new().read(true)).read_to_cleared_string(dst)
    }

    /// Read the entire contents of a file to the given buffer.
    #[track_caller]
    pub fn read_append_to_string<'dst>(&mut self, dst: &'dst mut String) -> &'dst mut String {
        match self.inner.read_to_string(dst) {
            Ok(_) => {},
            Err(e) => panic_io(&e, "reading", self.path),
        }
        dst
    }

    #[track_caller]
    pub fn read_to_cleared_string<'dst>(&mut self, dst: &'dst mut String) -> &'dst mut String {
        dst.clear();
        self.read_append_to_string(dst)
    }

    /// Replaces the entire contents of a file.
    #[track_caller]
    pub fn replace_contents(&mut self, data: &[u8]) {
        let res = match self.inner.seek(SeekFrom::Start(0)) {
            Ok(_) => match self.inner.write_all(data) {
                Ok(()) => self.inner.set_len(data.len() as u64),
                Err(e) => Err(e),
            },
            Err(e) => Err(e),
        };
        if let Err(e) = res {
            panic_io(&e, "writing", self.path);
        }
    }
}

/// Returns the path to the `cargo-clippy` binary
///
/// # Panics
///
/// Panics if the path of current executable could not be retrieved.
#[must_use]
pub fn cargo_clippy_path() -> PathBuf {
    let mut path = env::current_exe().expect("failed to get current executable name");
    path.set_file_name(CARGO_CLIPPY_EXE);
    path
}

#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub struct Version {
    pub major: u16,
    pub minor: u16,
}
impl FromStr for Version {
    type Err = ();
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if let Some(s) = s.strip_prefix("0.")
            && let Some((major, minor)) = s.split_once('.')
            && let Ok(major) = major.parse()
            && let Ok(minor) = minor.parse()
        {
            Ok(Self { major, minor })
        } else {
            Err(())
        }
    }
}
impl Version {
    /// Displays the version as a rust version. i.e. `x.y.0`
    #[must_use]
    pub fn rust_display(self) -> impl Display {
        struct X(Version);
        impl Display for X {
            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                write!(f, "{}.{}.0", self.0.major, self.0.minor)
            }
        }
        X(self)
    }

    /// Displays the version as it should appear in clippy's toml files. i.e. `0.x.y`
    #[must_use]
    pub fn toml_display(self) -> impl Display {
        struct X(Version);
        impl Display for X {
            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                write!(f, "0.{}.{}", self.0.major, self.0.minor)
            }
        }
        X(self)
    }
}

pub struct ClippyInfo {
    pub path: PathBuf,
    pub version: Version,
}
impl ClippyInfo {
    #[must_use]
    pub fn search_for_manifest() -> Self {
        let mut path = env::current_dir().expect("error reading the working directory");
        let mut buf = String::new();
        loop {
            path.push("Cargo.toml");
            if let Some(mut file) = File::open_if_exists(&path, OpenOptions::new().read(true)) {
                let mut in_package = false;
                let mut is_clippy = false;
                let mut version: Option<Version> = None;

                // Ad-hoc parsing to avoid dependencies. We control all the file so this
                // isn't actually a problem
                for line in file.read_to_cleared_string(&mut buf).lines() {
                    if line.starts_with('[') {
                        in_package = line.starts_with("[package]");
                    } else if in_package && let Some((name, value)) = line.split_once('=') {
                        match name.trim() {
                            "name" => is_clippy = value.trim() == "\"clippy\"",
                            "version"
                                if let Some(value) = value.trim().strip_prefix('"')
                                    && let Some(value) = value.strip_suffix('"') =>
                            {
                                version = value.parse().ok();
                            },
                            _ => {},
                        }
                    }
                }

                if is_clippy {
                    let Some(version) = version else {
                        panic!("error reading clippy version from {}", file.path.display());
                    };
                    path.pop();
                    return ClippyInfo { path, version };
                }
            }

            path.pop();
            assert!(
                path.pop(),
                "error finding project root, please run from inside the clippy directory"
            );
        }
    }
}

/// # Panics
/// Panics if given command result was failed.
pub fn exit_if_err(status: io::Result<ExitStatus>) {
    match status.expect("failed to run command").code() {
        Some(0) => {},
        Some(n) => process::exit(n),
        None => {
            eprintln!("Killed by signal");
            process::exit(1);
        },
    }
}

#[derive(Clone, Copy, PartialEq, Eq)]
pub enum UpdateMode {
    Check,
    Change,
}

pub(crate) fn exit_with_failure() {
    println!(
        "Not all lints defined properly. \
                 Please run `cargo dev update_lints` to make sure all lints are defined properly."
    );
    process::exit(1);
}

/// Replaces a region in a file delimited by two lines matching regexes.
///
/// `path` is the relative path to the file on which you want to perform the replacement.
///
/// See `replace_region_in_text` for documentation of the other options.
///
/// # Panics
///
/// Panics if the path could not read or then written
pub(crate) fn replace_region_in_file(
    update_mode: UpdateMode,
    path: &Path,
    start: &str,
    end: &str,
    write_replacement: impl FnMut(&mut String),
) {
    let contents = fs::read_to_string(path).unwrap_or_else(|e| panic!("Cannot read from `{}`: {e}", path.display()));
    let new_contents = match replace_region_in_text(&contents, start, end, write_replacement) {
        Ok(x) => x,
        Err(delim) => panic!("Couldn't find `{delim}` in file `{}`", path.display()),
    };

    match update_mode {
        UpdateMode::Check if contents != new_contents => exit_with_failure(),
        UpdateMode::Check => (),
        UpdateMode::Change => {
            if let Err(e) = fs::write(path, new_contents.as_bytes()) {
                panic!("Cannot write to `{}`: {e}", path.display());
            }
        },
    }
}

/// Replaces a region in a text delimited by two strings. Returns the new text if both delimiters
/// were found, or the missing delimiter if not.
pub(crate) fn replace_region_in_text<'a>(
    text: &str,
    start: &'a str,
    end: &'a str,
    mut write_replacement: impl FnMut(&mut String),
) -> Result<String, &'a str> {
    let (text_start, rest) = text.split_once(start).ok_or(start)?;
    let (_, text_end) = rest.split_once(end).ok_or(end)?;

    let mut res = String::with_capacity(text.len() + 4096);
    res.push_str(text_start);
    res.push_str(start);
    write_replacement(&mut res);
    res.push_str(end);
    res.push_str(text_end);

    Ok(res)
}
