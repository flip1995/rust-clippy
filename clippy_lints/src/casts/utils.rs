use rustc_middle::mir::interpret::{ConstValue, Scalar};
use rustc_middle::ty::{self, AdtDef, IntTy, Ty, TyCtxt, UintTy, VariantDiscr};
use rustc_target::abi::Size;

/// Returns the size in bits of an integral type.
/// Will return 0 if the type is not an int or uint variant
pub(super) fn int_ty_to_nbits(typ: Ty<'_>, tcx: TyCtxt<'_>) -> u64 {
    match typ.kind() {
        ty::Int(i) => match i {
            IntTy::Isize => tcx.data_layout.pointer_size.bits(),
            IntTy::I8 => 8,
            IntTy::I16 => 16,
            IntTy::I32 => 32,
            IntTy::I64 => 64,
            IntTy::I128 => 128,
        },
        ty::Uint(i) => match i {
            UintTy::Usize => tcx.data_layout.pointer_size.bits(),
            UintTy::U8 => 8,
            UintTy::U16 => 16,
            UintTy::U32 => 32,
            UintTy::U64 => 64,
            UintTy::U128 => 128,
        },
        _ => 0,
    }
}

pub(super) fn enum_ty_to_nbits(adt: &AdtDef, tcx: TyCtxt<'_>) -> u64 {
    let mut explicit = 0i128;
    let (start, end) = adt
        .variants
        .iter()
        .fold((i128::MAX, i128::MIN), |(start, end), variant| match variant.discr {
            VariantDiscr::Relative(x) => match explicit.checked_add(i128::from(x)) {
                Some(x) => (start, end.max(x)),
                None => (i128::MIN, end),
            },
            VariantDiscr::Explicit(id) => {
                let ty = tcx.type_of(id);
                if let Ok(ConstValue::Scalar(Scalar::Int(value))) = tcx.const_eval_poly(id) {
                    #[allow(clippy::cast_possible_truncation, clippy::cast_possible_wrap)]
                    let value = match (value.size().bytes(), ty.kind()) {
                        (1, ty::Int(_)) => i128::from(value.assert_bits(Size::from_bytes(1)) as u8 as i8),
                        (1, ty::Uint(_)) => i128::from(value.assert_bits(Size::from_bytes(1)) as u8),
                        (2, ty::Int(_)) => i128::from(value.assert_bits(Size::from_bytes(2)) as u16 as i16),
                        (2, ty::Uint(_)) => i128::from(value.assert_bits(Size::from_bytes(2)) as u16),
                        (4, ty::Int(_)) => i128::from(value.assert_bits(Size::from_bytes(4)) as u32 as i32),
                        (4, ty::Uint(_)) => i128::from(value.assert_bits(Size::from_bytes(4)) as u32),
                        (8, ty::Int(_)) => i128::from(value.assert_bits(Size::from_bytes(8)) as u64 as i64),
                        (8, ty::Uint(_)) => i128::from(value.assert_bits(Size::from_bytes(8)) as u64),
                        (16, ty::Int(_)) => value.assert_bits(Size::from_bytes(16)) as i128,
                        (16, ty::Uint(_)) => match i128::try_from(value.assert_bits(Size::from_bytes(16))) {
                            Ok(x) => x,
                            // Requires 128 bits
                            Err(_) => return (i128::MIN, end),
                        },
                        // Shouldn't happen if compilation was successful
                        _ => return (start, end),
                    };
                    explicit = value;
                    (start.min(value), end.max(value))
                } else {
                    // Shouldn't happen if compilation was successful
                    (start, end)
                }
            },
        });

    if start >= end {
        0
    } else {
        let neg_bits = if start < 0 {
            128 - (-(start + 1)).leading_zeros() + 1
        } else {
            0
        };
        let pos_bits = if end > 0 { 128 - end.leading_zeros() } else { 0 };
        neg_bits.max(pos_bits).into()
    }
}
