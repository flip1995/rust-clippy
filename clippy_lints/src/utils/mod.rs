pub mod author;
pub mod dump_hir;
pub mod format_args_collector;
#[cfg(feature = "internal")]
pub mod internal_lints;

// ==================================================================
// Configuration
// ==================================================================

// Shamelessly stolen from find_all (https://github.com/nectariner/find_all)
pub trait FindAll: Iterator + Sized {
    fn find_all<P>(&mut self, predicate: P) -> Option<Vec<usize>>
    where
        P: FnMut(&Self::Item) -> bool;
}

impl<I> FindAll for I
where
    I: Iterator,
{
    fn find_all<P>(&mut self, mut predicate: P) -> Option<Vec<usize>>
    where
        P: FnMut(&Self::Item) -> bool,
    {
        let mut occurences = Vec::<usize>::default();
        for (index, element) in self.enumerate() {
            if predicate(&element) {
                occurences.push(index);
            }
        }

        match occurences.len() {
            0 => None,
            _ => Some(occurences),
        }
    }
}
