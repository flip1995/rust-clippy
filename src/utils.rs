use rustc::lint::{Context, Lint};
use syntax::ast::{DefId, Name, Path};
use syntax::codemap::{ExpnInfo, Span};
use syntax::ptr::P;
use rustc::middle::ty;
use std::borrow::{Cow, IntoCow};
use std::convert::From;

/// returns true if the macro that expanded the crate was outside of
/// the current crate or was a compiler plugin
pub fn in_macro(cx: &Context, opt_info: Option<&ExpnInfo>) -> bool {
	// no ExpnInfo = no macro
	opt_info.map_or(false, |info| {
		// no span for the callee = external macro
		info.callee.span.map_or(true, |span| {
			// no snippet = external macro or compiler-builtin expansion
			cx.sess().codemap().span_to_snippet(span).ok().map_or(true, |code| 
				// macro doesn't start with "macro_rules"
				// = compiler plugin
				!code.starts_with("macro_rules")
			)
		})
	})
}

/// invokes in_macro with the expansion info of the given span
pub fn in_external_macro(cx: &Context, span: Span) -> bool {
	cx.sess().codemap().with_expn_info(span.expn_id, 
			|info| in_macro(cx, info))
}

/// check if a DefId's path matches the given absolute type path
/// usage e.g. with
/// `match_def_path(cx, id, &["core", "option", "Option"])`
pub fn match_def_path(cx: &Context, def_id: DefId, path: &[&str]) -> bool {
	cx.tcx.with_path(def_id, |iter| iter.map(|elem| elem.name())
		.zip(path.iter()).all(|(nm, p)| &nm.as_str() == p))
}

/// match a Path against a slice of segment string literals, e.g.
/// `match_path(path, &["std", "rt", "begin_unwind"])`
pub fn match_path(path: &Path, segments: &[&str]) -> bool {
	path.segments.iter().rev().zip(segments.iter().rev()).all(
		|(a,b)| a.identifier.as_str() == *b)
}

/// convert a span to a code snippet if available, otherwise use default, e.g.
/// `snippet(cx, expr.span, "..")`
pub fn snippet<'a>(cx: &Context, span: Span, default: &'a str) -> Cow<'a, str> {
	cx.sess().codemap().span_to_snippet(span).map(From::from).unwrap_or(Cow::Borrowed(default))
}

/// dereference a P<T> and return a ref on the result
pub fn de_p<T>(p: &P<T>) -> &T { &*p }

#[cfg(not(feature="structured_logging"))]
pub fn span_lint(cx: &Context, lint: &'static Lint, sp: Span, msg: &str) {
	cx.span_lint(lint, sp, msg);
}

#[cfg(feature="structured_logging")]
pub fn span_lint(cx: &Context, lint: &'static Lint, sp: Span, msg: &str) {
	// lint.name / lint.desc is can give details of the lint
	// cx.sess().codemap() has all these nice functions for line/column/snippet details
	// http://doc.rust-lang.org/syntax/codemap/struct.CodeMap.html#method.span_to_string
	cx.span_lint(lint, sp, msg);
}
