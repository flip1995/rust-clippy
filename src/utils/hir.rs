use consts::constant;
use rustc::lint::*;
use rustc_front::hir::*;
use syntax::ptr::P;

/// Type used to check whether two ast are the same. This is different from the operator
/// `==` on ast types as this operator would compare true equality with ID and span.
///
/// Note that some expressions kinds are not considered but could be added.
pub struct SpanlessEq<'a, 'tcx: 'a> {
    /// Context used to evaluate constant expressions.
    cx: &'a LateContext<'a, 'tcx>,
    /// If is true, never consider as equal expressions containing fonction calls.
    ignore_fn: bool,
}

impl<'a, 'tcx: 'a> SpanlessEq<'a, 'tcx> {
    pub fn new(cx: &'a LateContext<'a, 'tcx>) -> Self {
        SpanlessEq { cx: cx, ignore_fn: false }
    }

    pub fn ignore_fn(self) -> Self {
        SpanlessEq { cx: self.cx, ignore_fn: true }
    }

    /// Check whether two statements are the same.
    pub fn eq_stmt(&self, left: &Stmt, right: &Stmt) -> bool {
        match (&left.node, &right.node) {
            (&StmtDecl(ref l, _), &StmtDecl(ref r, _)) => {
                if let (&DeclLocal(ref l), &DeclLocal(ref r)) = (&l.node, &r.node) {
                    // TODO: tys
                    l.ty.is_none() && r.ty.is_none() &&
                        both(&l.init, &r.init, |l, r| self.eq_expr(l, r))
                }
                else {
                    false
                }
            }
            (&StmtExpr(ref l, _), &StmtExpr(ref r, _)) => self.eq_expr(l, r),
            (&StmtSemi(ref l, _), &StmtSemi(ref r, _)) => self.eq_expr(l, r),
            _ => false,
        }
    }

    /// Check whether two blocks are the same.
    pub fn eq_block(&self, left: &Block, right: &Block) -> bool {
        over(&left.stmts, &right.stmts, |l, r| self.eq_stmt(l, r)) &&
            both(&left.expr, &right.expr, |l, r| self.eq_expr(l, r))
    }

    // ok, itâ€™s a big function, but mostly one big match with simples cases
    #[allow(cyclomatic_complexity)]
    pub fn eq_expr(&self, left: &Expr, right: &Expr) -> bool {
        if let (Some(l), Some(r)) = (constant(self.cx, left), constant(self.cx, right)) {
            if l == r {
                return true;
            }
        }

        match (&left.node, &right.node) {
            (&ExprAddrOf(ref lmut, ref le), &ExprAddrOf(ref rmut, ref re)) => {
                lmut == rmut && self.eq_expr(le, re)
            }
            (&ExprAgain(li), &ExprAgain(ri)) => {
                both(&li, &ri, |l, r| l.node.name.as_str() == r.node.name.as_str())
            }
            (&ExprAssign(ref ll, ref lr), &ExprAssign(ref rl, ref rr)) => {
                self.eq_expr(ll, rl) && self.eq_expr(lr, rr)
            }
            (&ExprAssignOp(ref lo, ref ll, ref lr), &ExprAssignOp(ref ro, ref rl, ref rr)) => {
                lo.node == ro.node && self.eq_expr(ll, rl) && self.eq_expr(lr, rr)
            }
            (&ExprBlock(ref l), &ExprBlock(ref r)) => {
                self.eq_block(l, r)
            }
            (&ExprBinary(lop, ref ll, ref lr), &ExprBinary(rop, ref rl, ref rr)) => {
                lop.node == rop.node && self.eq_expr(ll, rl) && self.eq_expr(lr, rr)
            }
            (&ExprBreak(li), &ExprBreak(ri)) => {
                both(&li, &ri, |l, r| l.node.name.as_str() == r.node.name.as_str())
            }
            (&ExprBox(ref l), &ExprBox(ref r)) => {
                self.eq_expr(l, r)
            }
            (&ExprCall(ref lfun, ref largs), &ExprCall(ref rfun, ref rargs)) => {
                !self.ignore_fn &&
                    self.eq_expr(lfun, rfun) &&
                    self.eq_exprs(largs, rargs)
            }
            (&ExprCast(ref lx, ref lt), &ExprCast(ref rx, ref rt)) => {
                self.eq_expr(lx, rx) && self.eq_ty(lt, rt)
            }
            (&ExprField(ref lfexp, ref lfident), &ExprField(ref rfexp, ref rfident)) => {
                lfident.node == rfident.node && self.eq_expr(lfexp, rfexp)
            }
            (&ExprIndex(ref la, ref li), &ExprIndex(ref ra, ref ri)) => {
                self.eq_expr(la, ra) && self.eq_expr(li, ri)
            }
            (&ExprIf(ref lc, ref lt, ref le), &ExprIf(ref rc, ref rt, ref re)) => {
                self.eq_expr(lc, rc) &&
                    self.eq_block(lt, rt) &&
                    both(le, re, |l, r| self.eq_expr(l, r))
            }
            (&ExprLit(ref l), &ExprLit(ref r)) => l.node == r.node,
            (&ExprMatch(ref le, ref la, ref ls), &ExprMatch(ref re, ref ra, ref rs)) => {
                ls == rs &&
                    self.eq_expr(le, re) &&
                    over(la, ra, |l, r| {
                        self.eq_expr(&l.body, &r.body) &&
                            both(&l.guard, &r.guard, |l, r| self.eq_expr(l, r)) &&
                            over(&l.pats, &r.pats, |l, r| self.eq_pat(l, r))
                    })
            }
            (&ExprMethodCall(ref lname, ref ltys, ref largs), &ExprMethodCall(ref rname, ref rtys, ref rargs)) => {
                // TODO: tys
                !self.ignore_fn &&
                    lname.node == rname.node &&
                    ltys.is_empty() &&
                    rtys.is_empty() &&
                    self.eq_exprs(largs, rargs)
            }
            (&ExprRange(ref lb, ref le), &ExprRange(ref rb, ref re)) => {
                both(lb, rb, |l, r| self.eq_expr(l, r)) &&
                both(le, re, |l, r| self.eq_expr(l, r))
            }
            (&ExprRepeat(ref le, ref ll), &ExprRepeat(ref re, ref rl)) => {
                self.eq_expr(le, re) && self.eq_expr(ll, rl)
            }
            (&ExprRet(ref l), &ExprRet(ref r)) => {
                both(l, r, |l, r| self.eq_expr(l, r))
            }
            (&ExprPath(ref lqself, ref lsubpath), &ExprPath(ref rqself, ref rsubpath)) => {
                both(lqself, rqself, |l, r| self.eq_qself(l, r)) && self.eq_path(lsubpath, rsubpath)
            }
            (&ExprTup(ref ltup), &ExprTup(ref rtup)) => self.eq_exprs(ltup, rtup),
            (&ExprTupField(ref le, li), &ExprTupField(ref re, ri)) => {
                li.node == ri.node && self.eq_expr(le, re)
            }
            (&ExprUnary(lop, ref le), &ExprUnary(rop, ref re)) => {
                lop == rop && self.eq_expr(le, re)
            }
            (&ExprVec(ref l), &ExprVec(ref r)) => self.eq_exprs(l, r),
            (&ExprWhile(ref lc, ref lb, ref ll), &ExprWhile(ref rc, ref rb, ref rl)) => {
                self.eq_expr(lc, rc) &&
                    self.eq_block(lb, rb) &&
                    both(ll, rl, |l, r| l.name.as_str() == r.name.as_str())
            }
            _ => false,
        }
    }

    fn eq_exprs(&self, left: &[P<Expr>], right: &[P<Expr>]) -> bool {
        over(left, right, |l, r| self.eq_expr(l, r))
    }

    /// Check whether two patterns are the same.
    pub fn eq_pat(&self, left: &Pat, right: &Pat) -> bool {
        match (&left.node, &right.node) {
            (&PatBox(ref l), &PatBox(ref r)) => {
                self.eq_pat(l, r)
            }
            (&PatEnum(ref lp, ref la), &PatEnum(ref rp, ref ra)) => {
                self.eq_path(lp, rp) &&
                    both(la, ra, |l, r| {
                        over(l, r, |l, r| self.eq_pat(l, r))
                    })
            }
            (&PatIdent(ref lb, ref li, ref lp), &PatIdent(ref rb, ref ri, ref rp)) => {
                lb == rb && li.node.name.as_str() == ri.node.name.as_str() &&
                    both(lp, rp, |l, r| self.eq_pat(l, r))
            }
            (&PatLit(ref l), &PatLit(ref r)) => {
                self.eq_expr(l, r)
            }
            (&PatQPath(ref ls, ref lp), &PatQPath(ref rs, ref rp)) => {
                self.eq_qself(ls, rs) && self.eq_path(lp, rp)
            }
            (&PatTup(ref l), &PatTup(ref r)) => {
                over(l, r, |l, r| self.eq_pat(l, r))
            }
            (&PatRange(ref ls, ref le), &PatRange(ref rs, ref re)) => {
                self.eq_expr(ls, rs) &&
                    self.eq_expr(le, re)
            }
            (&PatRegion(ref le, ref lm), &PatRegion(ref re, ref rm)) => {
                lm == rm && self.eq_pat(le, re)
            }
            (&PatVec(ref ls, ref li, ref le), &PatVec(ref rs, ref ri, ref re)) => {
                over(ls, rs, |l, r| self.eq_pat(l, r)) &&
                    over(le, re, |l, r| self.eq_pat(l, r)) &&
                    both(li, ri, |l, r| self.eq_pat(l, r))
            }
            (&PatWild, &PatWild) => true,
            _ => false,
        }
    }

    fn eq_path(&self, left: &Path, right: &Path) -> bool {
        // The == of idents doesn't work with different contexts,
        // we have to be explicit about hygiene
        left.global == right.global &&
        over(&left.segments,
             &right.segments,
             |l, r| l.identifier.name.as_str() == r.identifier.name.as_str() && l.parameters == r.parameters)
    }

    fn eq_qself(&self, left: &QSelf, right: &QSelf) -> bool {
        left.ty.node == right.ty.node && left.position == right.position
    }

    fn eq_ty(&self, left: &Ty, right: &Ty) -> bool {
        match (&left.node, &right.node) {
            (&TyVec(ref lvec), &TyVec(ref rvec)) => self.eq_ty(lvec, rvec),
            (&TyPtr(ref lmut), &TyPtr(ref rmut)) => lmut.mutbl == rmut.mutbl && self.eq_ty(&*lmut.ty, &*rmut.ty),
            (&TyRptr(_, ref lrmut), &TyRptr(_, ref rrmut)) => {
                lrmut.mutbl == rrmut.mutbl && self.eq_ty(&*lrmut.ty, &*rrmut.ty)
            }
            (&TyPath(ref lq, ref lpath), &TyPath(ref rq, ref rpath)) => {
                both(lq, rq, |l, r| self.eq_qself(l, r)) && self.eq_path(lpath, rpath)
            }
            (&TyInfer, &TyInfer) => true,
            _ => false,
        }
    }
}

/// Check if the two `Option`s are both `None` or some equal values as per `eq_fn`.
fn both<X, F>(l: &Option<X>, r: &Option<X>, mut eq_fn: F) -> bool
    where F: FnMut(&X, &X) -> bool
{
    l.as_ref().map_or_else(|| r.is_none(), |x| r.as_ref().map_or(false, |y| eq_fn(x, y)))
}

/// Check if two slices are equal as per `eq_fn`.
fn over<X, F>(left: &[X], right: &[X], mut eq_fn: F) -> bool
    where F: FnMut(&X, &X) -> bool
{
    left.len() == right.len() && left.iter().zip(right).all(|(x, y)| eq_fn(x, y))
}
