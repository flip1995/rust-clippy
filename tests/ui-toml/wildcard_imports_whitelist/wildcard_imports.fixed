#![warn(clippy::wildcard_imports)]

mod utils {
    pub fn print() {}
}

mod utils_plus {
    pub fn do_something() {}
}

use utils::*;
use utils_plus::do_something;
//~^ ERROR: usage of wildcard import

fn main() {
    print();
    do_something();
}
