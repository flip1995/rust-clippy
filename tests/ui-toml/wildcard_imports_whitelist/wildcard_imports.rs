#![warn(clippy::wildcard_imports)]

mod utils {
    pub fn print() {}
}

mod utils_plus {
    pub fn do_something() {}
}

use utils::*;
use utils_plus::*;
//~^ ERROR: usage of wildcard import

fn main() {
    print();
    do_something();
}
