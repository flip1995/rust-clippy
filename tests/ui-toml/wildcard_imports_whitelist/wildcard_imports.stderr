error: usage of wildcard import
  --> $DIR/wildcard_imports.rs:12:5
   |
LL | use utils_plus::*;
   |     ^^^^^^^^^^^^^ help: try: `utils_plus::do_something`
   |
   = note: `-D clippy::wildcard-imports` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::wildcard_imports)]`

error: aborting due to 1 previous error

